data/src/test/java/org/apache/iceberg/io/TestGenericSortedPosDeleteWriter.java

```diff
@@ -0,0 +1,315 @@
+
+  private final FileFormat format;
+
+  private OutputFileFactory fileFactory;
+  private Record gRecord;
+
+  @Parameterized.Parameters(name = "FileFormat={0}")
+  public static Object[] parameters() {
+    return new Object[][] {
+        new Object[] {"avro"},
+        new Object[] {"parquet"},
+    };
+  }
+
+  public TestGenericSortedPosDeleteWriter(String fileFormat) {
+    super(FORMAT_V2);
+    this.format = FileFormat.valueOf(fileFormat.toUpperCase(Locale.ENGLISH));
+  }
+
+  @Before
+  public void setupTable() throws IOException {
+    this.tableDir = temp.newFolder();
+    Assert.assertTrue(tableDir.delete());
+
+    this.metadataDir = new File(tableDir, "metadata");
+    this.table = create(SCHEMA, PartitionSpec.unpartitioned());
+    this.gRecord = GenericRecord.create(SCHEMA);
+
+    this.fileFactory = new OutputFileFactory(table.spec(), format, table.locationProvider(), table.io(),
+        table.encryption(), 1, 1);
+
+    table.updateProperties()
+        .defaultFormat(format)
+        .commit();
+  }
+
+  private EncryptedOutputFile createEncryptedOutputFile() {
+    return fileFactory.newOutputFile();
+  }
+
+  private DataFile prepareDataFile(FileAppenderFactory<Record> appenderFactory, List<Record> rowSet)
+      throws IOException {
+    DataWriter<Record> writer = appenderFactory.newDataWriter(createEncryptedOutputFile(), format, null);
+    try (DataWriter<Record> closeableWriter = writer) {
+      for (Record record : rowSet) {
+        closeableWriter.add(record);
+      }
+    }
+
+    return writer.toDataFile();
+  }
+
+  private Record createRow(Integer id, String data) {
+    Record row = gRecord.copy();
+    row.setField("id", id);
+    row.setField("data", data);
+    return row;
+  }
+
+  private StructLikeSet expectedRowSet(Iterable<Record> records) {
+    StructLikeSet set = StructLikeSet.create(table.schema().asStruct());
+    records.forEach(set::add);
+    return set;
+  }
+
+  private StructLikeSet actualRowSet(String... columns) throws IOException {
+    StructLikeSet set = StructLikeSet.create(table.schema().asStruct());
+    try (CloseableIterable<Record> reader = IcebergGenerics.read(table).select(columns).build()) {
+      reader.forEach(set::add);
+    }
+    return set;
+  }
+
+  @Test
+  public void testSortedPosDelete() throws IOException {
+    List<Record> rowSet = Lists.newArrayList(
+        createRow(0, "aaa"),
+        createRow(1, "bbb"),
+        createRow(2, "ccc"),
+        createRow(3, "ddd"),
+        createRow(4, "eee")
+    );
+
+    FileAppenderFactory<Record> appenderFactory = new GenericAppenderFactory(table.schema(), table.spec(),
+        null, null, null);
+    DataFile dataFile = prepareDataFile(appenderFactory, rowSet);
+
+    SortedPosDeleteWriter<Record> writer = new SortedPosDeleteWriter<>(appenderFactory, fileFactory, format, null, 100);
+    try (SortedPosDeleteWriter<Record> closeableWriter = writer) {
+      for (int index = 0; index < rowSet.size(); index += 2) {
```

Comment:
I think if we delete them in natural order, sorting them or not in delete writer will result in the correct order. Do we want to initialize the index as 4 and decrement the counter to test the sorting logic?